sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala

```diff
@@ -46,7 +46,18 @@ class RelationalGroupedDataset protected[sql](
 
   private[this] def toDF(aggExprs: Seq[Expression]): DataFrame = {
     val aggregates = if (df.sparkSession.sessionState.conf.dataFrameRetainGroupColumns) {
-      groupingExprs ++ aggExprs
+      val agg = Aggregate(groupingExprs, aggExprs.map(alias), df.logicalPlan)
+      val plan = df.sparkSession.sessionState.executePlan(agg).analyzed
+      val aggAnalyzed = plan.find(_.isInstanceOf[Aggregate]).getOrElse(agg).asInstanceOf[Aggregate]
```

Comment:
The same here. Can you put a comment to explain the top level node might not be an `Aggregate`